File: skiena_1_6_7.java
import java.util.*;

// this class represents the location of a chess piece or a square on a board
class Location
{
  int row;
  int column;

  Location(int rang, int colonne)
  {
    row = rang;
    column = colonne;
  }

  public String toString()
  {
    StringBuilder builder = new StringBuilder();
    builder.append("(").append(Integer.toString(row)).append(", ");
    builder.append(Integer.toString(column)).append(")");
    return builder.toString();
  }
}

// this class represents a chess board
class Board
{
  static final int MAX_SIDE = 8;
  String[] data = new String[MAX_SIDE];

  // this method checks to see if the current board contains 8 rows of "........"
  boolean is_empty()
  {
    for (String string : data)
      if (!string.equals("........"))
        return false;
    return true;
  }

  // this method finds a king (k or K) from board and returns the location of
  // the king
  Location find_king(char king)
  {
    int i, j = 0;
    for (i = 0; i < MAX_SIDE; i++)
    {
      j = data[i].indexOf(king);
      if (j != -1)
        break;
    }
    return new Location(i, j);
  }

  public String toString()
  {
    StringBuilder builder = new StringBuilder();
    for (String row : data)
      builder.append(row).append("\n");
    return builder.toString();
  }
}

// this class represents a blocker, whether it's a threat or a block (which is
// either a harmless hostile or a friendly). it contains information pertaining
// to the blocker such as its id (chess piece), its location and whether it's a
// threat or not.
class Blocker
{
  char id;
  Location location;
  boolean is_enemy;

  Blocker(char pi, Location loc, boolean enemy)
  {
    id = pi;
    location = loc;
    is_enemy = enemy;
  }

  public String toString()
  {
    StringBuilder builder = new StringBuilder();
    builder.append(is_enemy ? "checked" : "blocked");
    builder.append(" by ").append(id).append(" at ").append(location);
    return builder.toString();
  }
}

// this class represents a king on a chess board
class King
{
  Board board;
  Location location;

  King(Board bd, Location loc)
  {
    board = bd;
    location = loc;
  }

  // this method checks whether row and column are within bounds (0 and MAX_SIDE)
  // assuming that the increments have already been added to row and column.
  boolean condition(int row_increment, int column_increment, int row, int column)
  {
    boolean row_condition, column_condition;
    // whether row is positive and less than MAX_SIDE
    if (row_increment != 0)
      row_condition = (row < Board.MAX_SIDE && row >= 0);
    else
      row_condition = true;
    // whether column is positive and less than MAX_SIDE
    if (column_increment != 0)
      column_condition = (column < Board.MAX_SIDE && column >= 0);
    else
      column_condition = true;

    // whether both row and column are within bounds
    return row_condition && column_condition;
  }

  // this method scans in 8 directions: up, down, left, right, and the 4
  // diagonal directions for the first threat (an opponent Queen, Rook or Bishop)
  // or a block (a friendly or a harmless opponent which may or may not be in
  // the way of a threat). if a threat or a block is found, the blocker is
  // returned; the blocker contains information such as: the id of the chess
  // piece, its location on the chess board, and whether it's a threat or a
  // block. if no such threat or block is found, a null is returned.
  Blocker scan_8_directions(int row_increment, int column_increment, String straight_enemies)
  {
    Blocker blocker = null;
    // set a starting point for scanning before looping.
    int i = location.row + row_increment;
    int j = location.column + column_increment;
    // keep scanning until out of bounds or a blocker is met
    while (condition(row_increment, column_increment, i, j))
    {
      char square = board.data[i].charAt(j);
      // square is blocked by some piece, friendly or enemy
      if (square != '.')
      {
        // save information pertaining to the blocker and stop scanning
        boolean is_enemy = (straight_enemies.indexOf(square) != -1);
        blocker = new Blocker(square, new Location(i, j), is_enemy);
        break;
      }
      else
      {
        // the increments can be 1, -1, or 0. for example, if row_increment is 1
        // then the scan is downward/southbound; additionally if column_increment
        // is -1 then the scan is southwestbound, if column_increment is 0 then
        // the scan is straight southbound, and if column_increment is 1 then
        // the scan is southeastbound. on the other hand, if row_increment is -1
        // then the scan is upward/northbound, and so on.
        i += row_increment;
        j += column_increment;
      }
    }
    return blocker;
  }

  // this method scans for a pawn in 2 positions in front of the current king
  // and for a knight in 8 positions around the king. if a pawn or a knight is
  // found, such is returned. otherwise a null is returned.
  Blocker scan_pawn_knight(int row_increment, int column_increment, char piece)
  {
    Blocker blocker = null;
    int i = location.row + row_increment;
    int j = location.column + column_increment;
    if (condition(row_increment, column_increment, i, j))
    {
      char square = board.data[i].charAt(j);
      if (square == piece)
        blocker = new Blocker(square, new Location(i, j), true);
    }
    return blocker;
  }
}

class skiena_1_6_7
{
  static List<Board> input()
  {
    List<Board> list = new ArrayList<>();
    Scanner scanner = new Scanner(System.in);
    while (true)
    {
      Board board = new Board();
      // read a total of 8 lines and store them in a board
      for (int i = 0; i < Board.MAX_SIDE; i++)
        board.data[i] = scanner.nextLine();

      // exit loop if current board is empty (containing 8 lines of "........")
      if (board.is_empty())
        break;
      else
      {
        // retain the current board
        list.add(board);
        // discard a blank line before moving on to the next board
        scanner.nextLine();
      }
    }

    scanner.close();
    return list;
  }

  static void output(List<Board> list)
  {
    // go thru each board
    for (int l = 0; l < list.size(); l++)
    {
      Board board = list.get(l);
      // System.out.print("\nBOARD:\n" + board);

      // black king (k) at index 0 and white king (K) at index 1. the kings'
      // index will determine the placement of their correponding opponents (QR,
      // QB, P) and the pawn increments in the other variables right below
      char[] kings = { 'k', 'K' };
      // Queen and Rook are enemies in the straight directions (north, south,
      // east, west). Note the placements of the opponents (QR and qr)
      // corresponding to k and K
      String[] straight_enemies = { "QR", "qr"};
      // Queen and Bishop are enemies in the 4 diagonal directions
      String[] diagonal_enemies = { "QB", "qb"};
      char[] pawns = { 'P', 'p' };
      // since black sides reside at the top of the chess board, the black king
      // (k) is only concerned with pawns in the south direction (1, positive
      // increment). likewise, the white king (K) is only concerned with pawns
      // in the north direction (-1, negative increment)
      int[] pawn_increments = { 1, -1 };
      char[] knights = { 'N', 'n' };
      String[] names = { "black", "white" };
      // master list of all blockers of both kings.
      // Java doesn't allow array of List, so must declare the array as follows
      List<List<Blocker>> obstructors = new ArrayList<List<Blocker>>();
      // go thru both kings. for each king, scan for 12 possible blockers: 4 in
      // the straight directions (north, south, east, west), 4 in diagonal
      // directions, 2 knights and 2 pawns (in 8 positions)
      for (int i = 0; i < kings.length; i++)
      {
        Location location = board.find_king(kings[i]);
        // System.out.println(kings[i] + " is at " + location);

        King king = new King(board, location);
        // list of all blockers of the current king
        List<Blocker> blockers = new ArrayList<>();

        // scan up and down for opponent Queen and Rook
        blockers.add(king.scan_8_directions(1, 0, straight_enemies[i]));
        blockers.add(king.scan_8_directions(-1, 0, straight_enemies[i]));
        // scan left and right for opponent Queen and Rook
        blockers.add(king.scan_8_directions(0, 1, straight_enemies[i]));
        blockers.add(king.scan_8_directions(0, -1, straight_enemies[i]));

        // scan the 4 diagonal directions for opponent Queen and Bishop
        blockers.add(king.scan_8_directions(1, 1, diagonal_enemies[i]));
        blockers.add(king.scan_8_directions(1, -1, diagonal_enemies[i]));
        blockers.add(king.scan_8_directions(-1, 1, diagonal_enemies[i]));
        blockers.add(king.scan_8_directions(-1, -1, diagonal_enemies[i]));

        // scan for opponent Pawn
        blockers.add(king.scan_pawn_knight(pawn_increments[i], 1, pawns[i]));
        blockers.add(king.scan_pawn_knight(pawn_increments[i], -1, pawns[i]));

        // scan 8 possible positions for opponent Knight
        blockers.add(king.scan_pawn_knight(2, 1, knights[i]));
        blockers.add(king.scan_pawn_knight(2, -1, knights[i]));
        blockers.add(king.scan_pawn_knight(-2, 1, knights[i]));
        blockers.add(king.scan_pawn_knight(-2, -1, knights[i]));
        blockers.add(king.scan_pawn_knight(1, 2, knights[i]));
        blockers.add(king.scan_pawn_knight(1, -2, knights[i]));
        blockers.add(king.scan_pawn_knight(-1, 2, knights[i]));
        blockers.add(king.scan_pawn_knight(-1, -2, knights[i]));

        // report on the blockers for debugging
        /*
        for (Blocker blocker : blockers)
          if (blocker != null)
            System.out.println(kings[i] + " is " + blocker);
        */

        obstructors.add(blockers);
      } // for loop ({ 'k', 'K' })

      // report as per problem requirement
      boolean checked = false;
      System.out.print("Game #" + (l+1) + ": ");
      for (int i = 0; i < kings.length; i++)
      {
        for (Blocker blocker : obstructors.get(i))
        {
          if (blocker != null && blocker.is_enemy)
          {
            System.out.println(names[i] + " king is in check.");
            checked = true;
            break;
          }
        }
        // one king is in check. since there is "no configurations where both
        // kings are check", there's no need to examine the other king.
        if (checked)
          break;
      }
      if (!checked)
        System.out.println("no king is in check.");
    } // for loop(List<Board> list)
  } // output() method

  public static void main(String[] args)
  {
    List<Board> list = input();
    output(list);
  }
}

File: skiena_1_6_7.c
#include <stdio.h>
#include <string.h>
#include <unistd.h> /* optarg */
#include "defs.h"


int _debug = DEBUG_OFF;


typedef int bool;
#define false 0
#define true 1


struct Location
{
    int _row;
    int _column;
};
typedef struct Location location_t;


#define BLACK 0
#define WHITE 1

struct Board
{
    char       _list[8][9];
    location_t _kings[2];
};
typedef struct Board board_t;


/* Check to see if the current board is empty */
bool b_empty(board_t* board)
{
    int i;
    for (i = 0; i < 8; i++)
        if (strcmp(board->_list[i], "........") != 0)
            return false;
    return true;
}


/* Find the king of this side (BLACK or WHITE) from the current board and
 * record the location of the king
 */
void find_king(board_t* board, int side)
{
    static char kings[] = { 'k', 'K' };
    location_t* king = board->_kings+side;
    int i;
    for (i = 0; i < 8; i++)
    {
        char* search = strchr(board->_list[i], kings[side]);
        if (search != NULL)
        {
            king->_row = i;
            king->_column = search-board->_list[i];
            if (_debug >= DEBUG_COPIOUS)
                printf("Found kings[%d] at (%d, %d)\n",
                       side, king->_row, king->_column);
            break;
        }
    }
}


bool input(char* input_file, board_t* list, int* count)
{
    bool success;
    FILE* file = fopen(input_file, "r");
    if (file == NULL)
    {
        printf("Cannot open file %s\n", input_file);
        success = false;
    }
    else
    {
        *count = 0;
        while (true)
        {
            char line[MAX_LINE];
            int i;
            for (i = 0; i < 8; i++)
            {
                /* read a line from file, discard the trailing newline, then
                 * store the line in the current board
                 */
                fgets(line, MAX_LINE, file);
                line[strlen(line) - 1] = '\0';
                strcpy(list[*count]._list[i], line);
            }

            /* exit loop if last board was empty */
            if (b_empty(list + *count))
                break;

            /* discard the blank line before moving on to the next board */
            else
            {
                fgets(line, 80, file);
                (*count)++;
            }
        }

        fclose(file);
        success = true;

        if (_debug >= DEBUG_COPIOUS)
            printf("READ %d boards\n", *count);
    }

    return success;
}


/* OPEN    means the current square is empty
 * BLOCKED means the current square is occupied by a chess piece, but since the
 *         piece is either a friendly or an unharmful hostile, the current king
 *         is not under attacked, but cannot go further
 * CHECKED means the current square is occupied by a hostile piece that is
 *         putting the current king under attack or in check
 */
typedef int status_t;

#define OPEN    10
#define BLOCKED 11
#define CHECKED 12


/* TPAWN, etc stand for this pawn, etc (on the same side as this king)
 * OPAWN, etc stand for opposing pawn, etc (on the opposing side)
 */
#define TPAWN   0
#define TKNIGHT 1
#define TBISHOP 2
#define TROOK   3
#define TQUEEN  4
#define OPAWN   5
#define OKNIGHT 6
#define OBISHOP 7
#define OROOK   8
#define OQUEEN  9
#define OKING   10


struct Attacker
{
    status_t   _status;
    char       _id;
    location_t _location;
};
typedef struct Attacker attacker_t;

void a_set(attacker_t* attacker, char id, int row, int column)
{
    attacker->_status = CHECKED;
    attacker->_id = id;
    attacker->_location._row = row;
    attacker->_location._column = column;
}


/* This function checks to see if the current square is open to the current
 * king, or is occupied by a chess piece that is unharmful to the king, or is
 * a hostile ROOK, QUEEN, or KING which puts the king in check
 */
attacker_t go_straight(board_t* board, int side, int row, int column,
                       bool next_door)
{
    static char symbols[2][11] =
    {
        { 'p', 'n', 'b', 'r', 'q', 'P', 'N', 'B', 'R', 'Q', 'K' },
        { 'P', 'N', 'B', 'R', 'Q', 'p', 'n', 'b', 'r', 'q', 'k' },
    };

    attacker_t attacker;
    char piece = board->_list[row][column];
    if (piece == '.')
        attacker._status = OPEN;
    else if (piece == symbols[side][TPAWN] || piece == symbols[side][TKNIGHT] ||
             piece == symbols[side][TBISHOP] || piece == symbols[side][TROOK] ||
             piece == symbols[side][TQUEEN] || piece == symbols[side][OPAWN] ||
             piece == symbols[side][OKNIGHT] || piece == symbols[side][OBISHOP])
        attacker._status = BLOCKED;
    else if (piece == symbols[side][OROOK] || piece == symbols[side][OQUEEN])
        a_set(&attacker, piece, row, column);
    else if (piece == symbols[side][OKING])
    {
        if (next_door)
            a_set(&attacker, piece, row, column);
        else
            attacker._status = BLOCKED;
    }

    if (_debug >= DEBUG_COPIOUS)
        printf("go_straight(%d, %d, %d, %d), piece %c, status %d\n",
               row, column, side, next_door, piece, attacker._status);
    return attacker;
}


/* This function checks to see if the current square is open to the current
 * king, or is occupied by a chess piece that is unharmful to the king, or is
 * a hostile BISHOP, QUEEN, or KING which puts the king in check
 */
attacker_t go_diagonal(board_t* board, int side, int row, int column,
                       bool next_door)
{
    static char symbols[2][11] =
    {
        { 'p', 'n', 'b', 'r', 'q', 'P', 'N', 'B', 'R', 'Q', 'K' },
        { 'P', 'N', 'B', 'R', 'Q', 'p', 'n', 'b', 'r', 'q', 'k' },
    };

    attacker_t attacker;
    char piece = board->_list[row][column];
    if (piece == '.')
        attacker._status = OPEN;
    else if (piece == symbols[side][TPAWN] || piece == symbols[side][TKNIGHT] ||
             piece == symbols[side][TBISHOP] || piece == symbols[side][TROOK] ||
             piece == symbols[side][TQUEEN] || piece == symbols[side][OPAWN] ||
             piece == symbols[side][OKNIGHT] || piece == symbols[side][OROOK])
        attacker._status = BLOCKED;
    else if (piece == symbols[side][OBISHOP] || piece == symbols[side][OQUEEN])
        a_set(&attacker, piece, row, column);
    else if (piece == symbols[side][OKING])
    {
        if (next_door)
            a_set(&attacker, piece, row, column);
        else
            attacker._status = BLOCKED;
    }

    if (_debug >= DEBUG_COPIOUS)
        printf("go_diagonal(%d, %d, %d, %d), piece %c, status %d\n",
               row, column, side, next_door, piece, attacker._status);
    return attacker;
}


#define MAX_ROWS    8
#define MAX_COLUMNS 8

/* This function checks to see if the current king is under check by a hostile
 * pawn by inspecting all 2 positions in the opposite direction where a pawn
 * might be
 */
attacker_t checked_pawn(board_t* board, int side)
{
    location_t* king = board->_kings+side;
    attacker_t attacker;
    if (side == BLACK)
    {
        if (king->_row+1 < MAX_ROWS && king->_column-1 >= 0 &&
            board->_list[king->_row+1][king->_column-1] == 'P')
            a_set(&attacker, 'P', king->_row+1, king->_column-1);
        else if (king->_row+1 < MAX_ROWS &&
                 king->_column+1 < MAX_COLUMNS &&
                 board->_list[king->_row+1][king->_column+1] == 'P')
            a_set(&attacker, 'P', king->_row+1, king->_column+1);
        else
            attacker._status = BLOCKED;
    }
    else
    {
        if (king->_row-1 >= 0 && king->_column-1 >= 0 &&
            board->_list[king->_row-1][king->_column-1] == 'p')
            a_set(&attacker, 'p', king->_row-1, king->_column-1);
        else if (king->_row-1 >= 0 &&
                 king->_column+1 < MAX_COLUMNS &&
                 board->_list[king->_row-1][king->_column+1] == 'p')
            a_set(&attacker, 'p', king->_row-1, king->_column+1);
        else
            attacker._status = BLOCKED;
    }

    if (_debug >= DEBUG_COPIOUS)
        printf("checked_pawn(%d), status %d\n", side, attacker._status);
    return attacker;
}


/* This function checks to see if the current king is under check by a hostile
 * knight by inspecting all 8 possible positions where a knight might be
 */
attacker_t checked_knight(board_t* board, int side)
{
    static char symbols[2] = { 'N', 'n' };
    location_t* king = board->_kings+side;
    attacker_t attacker;

    if (king->_row+2 < MAX_ROWS && king->_column-1 >= 0 &&
        board->_list[king->_row+2][king->_column-1] == symbols[side])
        a_set(&attacker, symbols[side], king->_row+2, king->_column-1);
    else if (king->_row+2 < MAX_ROWS && king->_column+1 < MAX_COLUMNS &&
        board->_list[king->_row+2][king->_column+1] == symbols[side])
        a_set(&attacker, symbols[side], king->_row+2, king->_column+1);
    else if (king->_row-2 >= 0 && king->_column-1 >= 0 &&
        board->_list[king->_row-2][king->_column-1] == symbols[side])
        a_set(&attacker, symbols[side], king->_row-2, king->_column-1);
    else if (king->_row-2 >= 0 && king->_column+1 < MAX_COLUMNS &&
        board->_list[king->_row-2][king->_column+1] == symbols[side])
        a_set(&attacker, symbols[side], king->_row-2, king->_column+1);
    else if (king->_row+1 < MAX_ROWS && king->_column-2 >= 0 &&
             board->_list[king->_row+1][king->_column-2] == symbols[side])
        a_set(&attacker, symbols[side], king->_row+1, king->_column-2);
    else if (king->_row+1 < MAX_ROWS && king->_column+2 < MAX_COLUMNS &&
             board->_list[king->_row+1][king->_column+2] == symbols[side])
        a_set(&attacker, symbols[side], king->_row+1, king->_column+2);
    else if (king->_row-1 >= 0 && king->_column-2 >= 0 &&
             board->_list[king->_row-1][king->_column-2] == symbols[side])
        a_set(&attacker, symbols[side], king->_row-1, king->_column-2);
    else if (king->_row-1 >= 0 && king->_column+2 < MAX_COLUMNS &&
             board->_list[king->_row-1][king->_column+2] == symbols[side])
        a_set(&attacker, symbols[side], king->_row-1, king->_column+2);
    else
        attacker._status = BLOCKED;

    if (_debug >= DEBUG_COPIOUS)
        printf("checked_knight(%d), status %d\n", side, attacker._status);
    return attacker;
}


/* This function checks to see if the current king is under check by a hostile
 * rook, queen, or king going north
 */
attacker_t checked_north(board_t* board, int side)
{
    location_t* king = board->_kings+side;
    attacker_t attacker;
    attacker._status = OPEN;
    int i;
    for (i = king->_row-1; i >= 0 && attacker._status == OPEN; i--)
        attacker = go_straight(board, side, i, king->_column,
                               (king->_row-i == 1));

    if (_debug >= DEBUG_COPIOUS)
        printf("checked_north(%d): status %d\n", side, attacker._status);
    return attacker;
}


/* This function checks to see if the current king is under check by a hostile
 * bishop, queen, or king going north-west
 */
attacker_t checked_north_east(board_t* board, int side)
{
    location_t* king = board->_kings+side;
    attacker_t attacker;
    attacker._status = OPEN;
    int i = king->_row-1;
    int j = king->_column+1;
    while (i >= 0 && j < 8 && attacker._status == OPEN)
    {
        attacker = go_diagonal(board, side, i, j,
                             (king->_row-i == 1 && j-king->_column == 1));
        i--;
        j++;
    }

    if (_debug >= DEBUG_COPIOUS)
        printf("checked_north_east(%d): status %d\n", side, attacker._status);
    return attacker;
}


attacker_t checked_east(board_t* board, int side)
{
    location_t* king = board->_kings+side;
    attacker_t attacker;
    attacker._status = OPEN;
    int j;
    for (j = king->_column+1; j < 8 && attacker._status == OPEN; j++)
        attacker = go_straight(board, side, king->_row, j,
                             (j-king->_column == 1));

    if (_debug >= DEBUG_COPIOUS)
        printf("checked_east(%d): status %d\n", side, attacker._status);
    return attacker;
}


attacker_t checked_south_east(board_t* board, int side)
{
    location_t* king = board->_kings+side;
    attacker_t attacker;
    attacker._status = OPEN;
    int i = king->_row+1;
    int j = king->_column+1;
    while (i < 8 && j < 8 && attacker._status == OPEN)
    {
        attacker = go_diagonal(board, side, i, j,
                             (i-king->_row == 1 && j-king->_column == 1));
        i++;
        j++;
    }

    if (_debug >= DEBUG_COPIOUS)
        printf("checked_south_east(%d): status %d\n", side, attacker._status);
    return attacker;
}


attacker_t checked_south(board_t* board, int side)
{
    location_t* king = board->_kings+side;
    attacker_t attacker;
    attacker._status = OPEN;
    int i;
    for (i = king->_row+1; i < 8 && attacker._status == OPEN; i++)
        attacker = go_straight(board, side, i, king->_column,
                             (i-king->_row == 1));

    if (_debug >= DEBUG_COPIOUS)
        printf("checked_south(%d): status %d\n", side, attacker._status);
    return attacker;
}


attacker_t checked_south_west(board_t* board, int side)
{
    location_t* king = board->_kings+side;
    attacker_t attacker;
    attacker._status = OPEN;
    int i = king->_row+1;
    int j = king->_column-1;
    while (i < 8 && j >= 0 && attacker._status == OPEN)
    {
        attacker = go_diagonal(board, side, i, j,
                             (i-king->_row == 1 && king->_column-j == 1));
        i++;
        j--;
    }

    if (_debug >= DEBUG_COPIOUS)
        printf("checked_south_west(%d): status %d\n", side, attacker._status);
    return attacker;
}


attacker_t checked_west(board_t* board, int side)
{
    location_t* king = board->_kings+side;
    attacker_t attacker;
    attacker._status = OPEN;
    int j;
    for (j = king->_column-1; j >= 0 && attacker._status == OPEN; j--)
        attacker = go_straight(board, side, king->_row, j,
                             (king->_column-j == 1));

    if (_debug >= DEBUG_COPIOUS)
        printf("checked_west(%d): status %d\n", side, attacker._status);
    return attacker;
}


attacker_t checked_north_west(board_t* board, int side)
{
    location_t* king = board->_kings+side;
    attacker_t attacker;
    attacker._status = OPEN;
    int i = king->_row-1;
    int j = king->_column-1;
    while (i >= 0 && j >= 0 && attacker._status == OPEN)
    {
        attacker = go_diagonal(board, side, i, j,
                             (king->_row-i == 1 && king->_column-j == 1));
        i--;
        j--;
    }

    if (_debug >= DEBUG_COPIOUS)
        printf("checked_north_west(%d): status %d\n", side, attacker._status);
    return attacker;
}


void output(board_t* boards, int count)
{
    if (_debug >= DEBUG_COPIOUS)
        printf("PROCESSING %d boards\n", count);

    char* kings[2] = { "black", "white" };
    int i;
    for (i = 0; i < count; i++)
    {
        /* print out current board */
        if (_debug >= DEBUG_LIGHT)
        {
            int j;
            for (j = 0; j < 8; j++)
                printf("%s\n", boards[i]._list[j]);
        }

        attacker_t attacker;
        int j;
        for (j = 0; j < 2; j++)
        {
            find_king(boards+i, j);

            /* From the current king's location, check:
             * (1) all 2 vertical and 2 horizontal directions for hostile
             *     rooks, queens and kings;
             * (2) all 4 diagonal directions for hostile bishops, queens and
             *     kings;
             * (3) all 8 positions for hostile knights; and
             * (4) all 2 positions for hostile pawns
             */
            attacker = checked_north(boards+i, j);
            if (attacker._status != CHECKED)
            {
                attacker = checked_north_east(boards+i, j);
                if (attacker._status != CHECKED)
                {
                    attacker = checked_east(boards+i, j);
                    if (attacker._status != CHECKED)
                    {
                        attacker = checked_south_east(boards+i, j);
                        if (attacker._status != CHECKED)
                        {
                            attacker = checked_south(boards+i, j);
                            if (attacker._status != CHECKED)
                            {
                                attacker = checked_south_west(boards+i, j);
                                if (attacker._status != CHECKED)
                                {
                                    attacker = checked_west(boards+i, j);
                                    if (attacker._status != CHECKED)
                                    {
                                        attacker = checked_north_west(boards+i, j);
                                        if (attacker._status != CHECKED)
                                        {
                                            attacker = checked_knight(boards+i, j);
                                            if (attacker._status != CHECKED)
                                                attacker = checked_pawn(boards+i, j);
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }

            location_t* king = boards[i]._kings+j;
            if (attacker._status == CHECKED)
            {
                if (_debug >= DEBUG_LIGHT)
                    printf("%c at (%d, %d) is in check by %c at (%d, %d)\n",
                           boards[i]._list[king->_row][king->_column],
                           king->_row, king->_column, attacker._id,
                           attacker._location._row, attacker._location._column);
                printf("Game #%d: %s king is in check.\n", i+1, kings[j]);
                break;
            }
            else
            {
                if (_debug >= DEBUG_LIGHT)
                    printf("%c at (%d, %d) is not in check.\n",
                           boards[i]._list[king->_row][king->_column],
                           king->_row, king->_column);
            }
        }

        if (attacker._status != CHECKED)
            printf("Game #%d: no king is in check.\n", i+1);
        if (_debug >= DEBUG_LIGHT)
            printf("\n");
    }
}


void main(int argc, char** argv)
{
    char opt;
    char* input_file = NULL;

    while ((opt = getopt(argc, argv, "i:d:")) != -1)
    {
        switch (opt)
        {
        case 'i':
            input_file = optarg;
            break;
        case 'd':
            _debug = atoi(optarg);
            break;
        default:
            printf("Usage: %s -i input_file [-d 0|1|2]\n", argv[0]);
        }
    }

    if (input_file == NULL || _debug < DEBUG_OFF || _debug > DEBUG_COPIOUS)
        printf("Usage: %s -i input_file [-d 0|1|2]\n", argv[0]);
    else
    {
        board_t list[MAX_CASES];
        int count;
        if (input(input_file, list, &count))
            output(list, count);
    }
}
